diff --git a/node_modules/@nomicfoundation/hardhat-viem/internal/clients.js b/node_modules/@nomicfoundation/hardhat-viem/internal/clients.js
index 9466df2..b3d2fd5 100644
--- a/node_modules/@nomicfoundation/hardhat-viem/internal/clients.js
+++ b/node_modules/@nomicfoundation/hardhat-viem/internal/clients.js
@@ -47,7 +47,7 @@ async function innerGetPublicClient(provider, chain, publicClientConfig) {
     const parameters = { ...defaultParameters, ...publicClientConfig };
     const publicClient = viem.createPublicClient({
         chain,
-        transport: viem.custom(provider),
+        transport: viem.custom(provider, { retryCount: 0 }),
         ...parameters,
     });
     return publicClient;
@@ -80,7 +80,7 @@ async function innerGetWalletClients(provider, chain, accounts, walletClientConf
     const walletClients = accounts.map((account) => viem.createWalletClient({
         chain,
         account,
-        transport: viem.custom(provider),
+        transport: viem.custom(provider, { retryCount: 0 }),
         ...parameters,
     }));
     return walletClients;
@@ -123,7 +123,7 @@ async function innerGetTestClient(provider, chain, mode, testClientConfig) {
     const testClient = viem.createTestClient({
         mode,
         chain,
-        transport: viem.custom(provider),
+        transport: viem.custom(provider, { retryCount: 0 }),
         ...parameters,
     });
     return testClient;
